Initial <input type=checkbox switch> animation support

annevk opened the pull request:
#### 0a64dd54421137c48a57e6e0aab15a99139a8776
<pre>
Initial &lt;input type=checkbox switch&gt; animation support
<a href="https://bugs.webkit.org/show_bug.cgi?id=265108">https://bugs.webkit.org/show_bug.cgi?id=265108</a>

Reviewed by Aditya Keerthi.

This adds the infrastructure needed to start the animation, somewhat
based on our &lt;progress&gt; implementation. And also implements the macOS
animation.

Thanks to Aditya, smfr, Tim Horton, Antoine, and Kimmo for much needed
guidance.

* Source/WebCore/dom/EventDispatcher.cpp:
(WebCore::EventDispatcher::dispatchEvent):
* Source/WebCore/html/CheckboxInputType.cpp:
(WebCore::CheckboxInputType::willDispatchClick):
(WebCore::switchCheckedChangeAnimationUpdateInterval):
(WebCore::CheckboxInputType::performSwitchCheckedChangeAnimation):
(WebCore::CheckboxInputType::switchCheckedChangeAnimationProgress const):
(WebCore::CheckboxInputType::switchCheckedChangeAnimationTimerFired):
* Source/WebCore/html/CheckboxInputType.h:
* Source/WebCore/html/HTMLInputElement.cpp:
(WebCore::HTMLInputElement::setChecked):
(WebCore::HTMLInputElement::switchCheckedChangeAnimationProgress const):
* Source/WebCore/html/HTMLInputElement.h:
* Source/WebCore/platform/graphics/controls/SwitchThumbPart.h:
* Source/WebCore/platform/graphics/controls/SwitchTrackPart.h:
* Source/WebCore/platform/graphics/mac/controls/SwitchMacUtilities.h:
* Source/WebCore/platform/graphics/mac/controls/SwitchMacUtilities.mm:
(WebCore::SwitchMacUtilities::easeInOut):
* Source/WebCore/platform/graphics/mac/controls/SwitchThumbMac.h:
* Source/WebCore/platform/graphics/mac/controls/SwitchThumbMac.mm:
(WebCore::SwitchThumbMac::cellSize const):
(WebCore::SwitchThumbMac::draw):
* Source/WebCore/platform/graphics/mac/controls/SwitchTrackMac.h:
* Source/WebCore/platform/graphics/mac/controls/SwitchTrackMac.mm:
(WebCore::trackMaskImage):
(WebCore::trackImage):
(WebCore::SwitchTrackMac::draw):
* Source/WebCore/rendering/RenderTheme.cpp:
(WebCore::updateSwitchThumbPartForRenderer):
(WebCore::updateSwitchTrackPartForRenderer):
(WebCore::RenderTheme::updateControlPartForRenderer const):
* Source/WebCore/rendering/RenderTheme.h:
(WebCore::RenderTheme::switchCheckedChangeAnimationDuration const):
* Source/WebCore/rendering/RenderThemeMac.h:
* Source/WebKit/Shared/WebCoreArgumentCoders.cpp:
(IPC::ArgumentCoder&lt;ControlPart&gt;::encode):
(IPC::ArgumentCoder&lt;ControlPart&gt;::decode):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Canonical link: <a href="https://commits.webkit.org/271118@main">https://commits.webkit.org/271118@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/134ca945fd9bf29657228ad611cd00985a70466d

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/27417 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/6057 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/28666 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/29644 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/25097 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/27886 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/7970 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/3454 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/24900 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/27680 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/4850 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/23538 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/4226 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/4382 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/24534 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/30280 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/25049 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/24956 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/30516 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/4396 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/2538 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/28475 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/5863 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/24269 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/6595 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/4853 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/4783 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

kkinnunen-apple commented on Source/WebCore/html/CheckboxInputType.cpp:
```diff
@@ -124,4 +130,47 @@ bool CheckboxInputType::shouldAppearIndeterminate() const
+{
+    ASSERT(isSwitch());
+    ASSERT(element());
+    ASSERT(element()->renderer());
```
Not the thing I was asked to comment about, but:
Ad hoc timers make me uneasy personally.
What if the system is supposed to animate at 30fps?
What if the system is supposed to animate at 120fps?
What if the element was removed from the tree and doesn't have a renderer?

I keep thinking about other active UI elements in WebKit, do they all have bespoke timers to run the animations?

kkinnunen-apple commented on Source/WebCore/platform/graphics/mac/controls/SwitchTrackMac.mm:
```diff
@@ -53,53 +52,50 @@
+    auto coreUIDirection = (__bridge NSString *)(isRTL ? kCUIUserInterfaceLayoutDirectionRightToLeft : kCUIUserInterfaceLayoutDirectionLeftToRight);
 
-    size.scale(style.zoomFactor);
+    GraphicsContextStateSaver stateSaver(context);
```
Conceptual and stylistic matter:
GraphicsContextStateSaver saves the GraphicsContext state.
Nothing here strictly mutates the GraphicsContext state.

You're smuggling the internal  image buffer context to the native drawing functions.
CGContextStateSaver would be more consistent usage here.

The most consistent would be to create a native (bitmap based) context and draw to that.
Then create CGImage out of that context and NativeImage out of that.

But this is just splitting hairs as the smuggling is done in so many places.

kkinnunen-apple commented on Source/WebCore/platform/graphics/mac/controls/SwitchTrackMac.mm:
```diff
@@ -53,53 +52,50 @@
-    }];
-
-    trackBuffer->context().clipToImageBuffer(*trackMaskBuffer, drawingTrackRect);
+    trackImage->context().clipToImageBuffer(*trackMaskImage, drawingTrackRect);
```
This clip especially is such a dubious smuggling.
Nothing says that if you clip in WebKit concept GraphicsContext, the clip would end up exactly correctly for you in the underlying platformContext() for the platformContext() to use.
It does today, maybe.

Not specifically problem of your patch, as this smuggling is pervasive in these parts of the code.
(One more reason to move stuff to WP, so this sort of thing is not possible at all)

pxlcoder commented on Source/WebCore/html/CheckboxInputType.h:
```diff
@@ -57,6 +60,12 @@ class CheckboxInputType final : public BaseCheckableInputType {
+    void switchAnimationTimerFired();
+
+    Seconds m_switchAnimationStartTime { 0_s };
+    Seconds m_switchAnimationDuration;
```
I don't think this needs to be a member variable.

annevk replied:
So we just call the theme method each time? That works. Progress does seem to store it though.

pxlcoder commented on Source/WebCore/html/HTMLInputElement.cpp:
```diff
@@ -1043,8 +1044,11 @@ void HTMLInputElement::setChecked(bool isChecked)
-    if (auto* renderer = this->renderer(); renderer && renderer->style().hasEffectiveAppearance())
+    if (auto* renderer = this->renderer(); renderer && renderer->style().hasEffectiveAppearance()) {
+        if (isSwitch())
+            downcast<CheckboxInputType>(*m_inputType).startSwitchAnimationTimer();
```
I would call the method something like `performSwitchStateChangeAnimation` (or something along those lines) rather than `startSwitchAnimationTimer`. The timer is really an implementation detail.

pxlcoder commented on Source/WebCore/platform/graphics/mac/controls/SwitchMacUtilities.mm:
```diff
@@ -77,6 +77,14 @@ static FloatRect rectForBounds(const FloatRect& bounds)
+{
+    // FIXME: this should be a MESSAGE_CHECK, ideally added to
+    // WebCoreArgumentCoders.serialization.in when it gains that capability.
+    auto clampedProgress = clampTo<float>(progress, 0.0f, 1.0f);
```
`easeInOut` is too far removed from the IPC layer to be worrying about validation.

Can't you add a `[Validator=...]` in `WebCoreArgumentCoders.serialization.in`?

If not, I would still move the clamping earlier, like in the constructor of the part.

pxlcoder commented on Source/WebCore/platform/graphics/mac/controls/SwitchThumbMac.mm:
```diff
@@ -76,9 +76,12 @@
 
     auto inflatedTrackRect = inflatedRect(trackRect, size, outsets, style);
 
+    auto drawingThumbSide = inflatedTrackRect.height();
```
I think you meant `size` rather than `side` here?

annevk replied:
I meant side as in each side has this length. A size would be a combination of a height and width in WebKit, I think.

pxlcoder replied:
Ah, I think `drawingThumbLength` would be better then.

pxlcoder commented on Source/WebCore/html/HTMLInputElement.cpp:
```diff
@@ -1043,8 +1044,11 @@ void HTMLInputElement::setChecked(bool isChecked)
-    if (auto* renderer = this->renderer(); renderer && renderer->style().hasEffectiveAppearance())
+    if (auto* renderer = this->renderer(); renderer && renderer->style().hasEffectiveAppearance()) {
+        if (isSwitch())
+            downcast<CheckboxInputType>(*m_inputType).performSwitchCheckedChangeAnimation();
```
Just realized that we shouldn't perform the animation when the checked state is changed programmatically.

(from testing on macOS, the animation only occurs when the user interacts with the control)

pxlcoder commented on Source/WebCore/rendering/RenderTheme.cpp:
```diff
@@ -601,6 +601,24 @@ static void updateSliderTrackPartForRenderer(SliderTrackPart& sliderTrackPart, c
+static void updateSwitchThumbPartForRenderer(SwitchThumbPart& switchThumbPart, const RenderObject& renderer)
+{
+    ASSERT(is<HTMLInputElement>(renderer.node()->shadowHost()));
+    auto& input = downcast<HTMLInputElement>(*renderer.node()->shadowHost());
```
I think this should be a `RELEASE_ASSERT`, since otherwise we have type confusion.

pxlcoder commented on Source/WebCore/rendering/RenderTheme.cpp:
```diff
@@ -601,6 +601,24 @@ static void updateSliderTrackPartForRenderer(SliderTrackPart& sliderTrackPart, c
+
+static void updateSwitchTrackPartForRenderer(SwitchTrackPart& switchTrackPart, const RenderObject& renderer)
+{
+    ASSERT(is<HTMLInputElement>(renderer.node()->shadowHost()));
```
Ditto.

annevk committed:
Initial <input type=checkbox switch> animation support
https://bugs.webkit.org/show_bug.cgi?id=265108

Reviewed by Aditya Keerthi.

This adds the infrastructure needed to start the animation, somewhat
based on our <progress> implementation. And also implements the macOS
animation.

Thanks to Aditya, smfr, Tim Horton, Antoine, and Kimmo for much needed
guidance.

* Source/WebCore/dom/EventDispatcher.cpp:
(WebCore::EventDispatcher::dispatchEvent):
* Source/WebCore/html/CheckboxInputType.cpp:
(WebCore::CheckboxInputType::willDispatchClick):
(WebCore::switchCheckedChangeAnimationUpdateInterval):
(WebCore::CheckboxInputType::performSwitchCheckedChangeAnimation):
(WebCore::CheckboxInputType::switchCheckedChangeAnimationProgress const):
(WebCore::CheckboxInputType::switchCheckedChangeAnimationTimerFired):
* Source/WebCore/html/CheckboxInputType.h:
* Source/WebCore/html/HTMLInputElement.cpp:
(WebCore::HTMLInputElement::setChecked):
(WebCore::HTMLInputElement::switchCheckedChangeAnimationProgress const):
* Source/WebCore/html/HTMLInputElement.h:
* Source/WebCore/platform/graphics/controls/SwitchThumbPart.h:
* Source/WebCore/platform/graphics/controls/SwitchTrackPart.h:
* Source/WebCore/platform/graphics/mac/controls/SwitchMacUtilities.h:
* Source/WebCore/platform/graphics/mac/controls/SwitchMacUtilities.mm:
(WebCore::SwitchMacUtilities::easeInOut):
* Source/WebCore/platform/graphics/mac/controls/SwitchThumbMac.h:
* Source/WebCore/platform/graphics/mac/controls/SwitchThumbMac.mm:
(WebCore::SwitchThumbMac::cellSize const):
(WebCore::SwitchThumbMac::draw):
* Source/WebCore/platform/graphics/mac/controls/SwitchTrackMac.h:
* Source/WebCore/platform/graphics/mac/controls/SwitchTrackMac.mm:
(WebCore::trackMaskImage):
(WebCore::trackImage):
(WebCore::SwitchTrackMac::draw):
* Source/WebCore/rendering/RenderTheme.cpp:
(WebCore::updateSwitchThumbPartForRenderer):
(WebCore::updateSwitchTrackPartForRenderer):
(WebCore::RenderTheme::updateControlPartForRenderer const):
* Source/WebCore/rendering/RenderTheme.h:
(WebCore::RenderTheme::switchCheckedChangeAnimationDuration const):
* Source/WebCore/rendering/RenderThemeMac.h:
* Source/WebKit/Shared/WebCoreArgumentCoders.cpp:
(IPC::ArgumentCoder<ControlPart>::encode):
(IPC::ArgumentCoder<ControlPart>::decode):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Canonical link: https://commits.webkit.org/271118@main
